src/MemoryTracing/MemoryInfoProducer.cpp

```diff
@@ -36,24 +40,66 @@ void MemoryInfoProducer::SetExitRequested(bool exit_requested) {
   exit_requested_ = exit_requested;
 }
 
-void MemoryInfoProducer::Run() {
-  orbit_base::SetCurrentThreadName("MemInfoPr::Run");
+MemoryInfoProducerRunFn SystemMemoryInfoProducerRunFn() {
+  return [](uint64_t sampling_period_ns, MemoryInfoListener* listener, int32_t /*pid*/,
+            bool* exit_requested, absl::Mutex* exit_requested_mutex) {
+    orbit_base::SetCurrentThreadName("SysMemPr::Run");
 
-  CHECK(listener_ != nullptr);
-  ProduceSystemMemoryUsageAndSendToListener();
+    CHECK(listener != nullptr);
+
+    absl::Time scheduled_time = absl::Now();
+    absl::MutexLock lock(exit_requested_mutex);
+    while (!(*exit_requested)) {
```

Comment:
there is some repitition going on maybe have the mutex and pereodic queries it done on the caller side
and have function only do GetSystemMemoryUsage and send it to the listener?